### PR TITLE
盘古BPM 批量查询支持

### PR DESCRIPTION
**Pull request by qiudaoke**
#### Check List:
* Unit tests: YES / NO / NA
* Documentation: YES / NO / NA

### CODE REVIEW

File: modules/flowable-engine/src/test/java/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.flowable.engine.test.Deployment;
+import org.flowable.engine.test.FlowableTestCase;
+import org.flowable.task.api.Task;
+import org.flowable.task.api.history.HistoricTaskInstance;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (dbmalkovsky) [inline modules/flowable-engine/src/test/java/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.java:24]: Thanks for the PR.

Please follow project coding standards for imports; see https://github.com/flowable/flowable-engine/tree/main/ide-settings.  

The standard order is:

- static imports
- java
- javax
- org
- com
- everything else

File: modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/HistoricTaskInstanceQueryImpl.java [hunk 2]
```diff
@@ -650,6 +652,34 @@ public HistoricTaskInstanceQuery taskName(String taskName) {
...
+            throw new FlowableIllegalArgumentException("Task id list is null");
+        }
+        if (taskIdList.isEmpty()) {
+            throw new FlowableIllegalArgumentException("Task id list is empty");
+        }
+        for (String taskId : taskIdList) {
+            if (taskId == null) {
+                throw new FlowableIllegalArgumentException("None of the given task ids can be null");
+            }
+        }
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/HistoricTaskInstanceQueryImpl.java:673]: I don't think that we need to do this, we are not doing it for any other collection

File: modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/HistoricTaskInstanceQueryImpl.java [hunk 3]
```diff
@@ -650,6 +652,34 @@ public HistoricTaskInstanceQuery taskName(String taskName) {
         return this;
     }
 
+    /**
+     * Only select tasks with a task id that is in the given list
+     *
+     * @param taskIdList
+     * @throws FlowableIllegalArgumentException When task id list is empty or <code>null</code> or contains <code>null String</code>.
+     */
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/HistoricTaskInstanceQueryImpl.java:660]: please move the javadoc to the interface only

File: modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/TaskQueryImpl.java [hunk 4]
```diff
@@ -200,6 +202,34 @@ public TaskQueryImpl taskName(String name) {
         return this;
     }
 
+    /**
+     * Only select tasks with a task id that is in the given list
+     *
+     * @param taskIdList
+     * @throws FlowableIllegalArgumentException When task id list is empty or <code>null</code> or contains <code>null String</code>.
+     */
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/TaskQueryImpl.java:210]: Please move the javadoc to the interface only

File: modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/TaskQueryImpl.java [hunk 5]
```diff
@@ -200,6 +202,34 @@ public TaskQueryImpl taskName(String name) {
...
+            throw new FlowableIllegalArgumentException("Task id list is null");
+        }
+        if (taskIdList.isEmpty()) {
+            throw new FlowableIllegalArgumentException("Task id list is empty");
+        }
+        for (String taskId : taskIdList) {
+            if (taskId == null) {
+                throw new FlowableIllegalArgumentException("None of the given task ids can be null");
+            }
+        }
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/java/org/flowable/task/service/impl/TaskQueryImpl.java:223]: Same comment as in the history query. We don't do this for any other collection, so we can remove it

File: modules/flowable-task-service-api/src/main/java/org/flowable/task/api/TaskInfoQuery.java [hunk 6]
```diff
@@ -789,4 +796,6 @@
      */
     T orderByCategory();
 
+
+
```
Comment (filiphr) [inline modules/flowable-task-service-api/src/main/java/org/flowable/task/api/TaskInfoQuery.java:800]: This is obsolete

File: modules/flowable-task-service/src/main/resources/org/flowable/task/service/db/mapping/entity/HistoricTaskInstance.xml [hunk 7]
```diff
@@ -996,6 +996,12 @@
       <if test="processInstanceId != null">
         and ${queryTablePrefix}PROC_INST_ID_ = #{processInstanceId}
       </if>
+    <if test="taskIdList != null &amp;&amp; taskIdList.size() &gt; 0">
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/resources/org/flowable/task/service/db/mapping/entity/HistoricTaskInstance.xml:999]: We recently did a change when checking for a size of a list. please use `!taskIdList.empty` instead.

File: modules/flowable-task-service/src/main/resources/org/flowable/task/service/db/mapping/entity/Task.xml [hunk 8]
```diff
@@ -556,6 +556,13 @@
       <if test="processInstanceId != null">
         and RES.PROC_INST_ID_ = #{processInstanceId}
       </if>
+      <if test="taskIdList != null &amp;&amp; taskIdList.size() &gt; 0">
+        and RES.ID_ IN
+        <foreach item="taskId" index="index" collection="taskIdList"
```
Comment (filiphr) [inline modules/flowable-task-service/src/main/resources/org/flowable/task/service/db/mapping/entity/Task.xml:561]: We recently did a change when checking for a size of a list. please use `!taskIdList.empty` instead.

File: modules/flowable-engine/src/test/resources/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.testSimpleProcess.bpmn20.xml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (filiphr) [inline modules/flowable-engine/src/test/resources/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.testSimpleProcess.bpmn20.xml:1]: This is not needed, you can reuse one of the current one task processes.

File: modules/flowable-engine/src/test/java/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import org.flowable.task.api.Task;
+import org.flowable.task.api.history.HistoricTaskInstance;
+/**
+ * @author 盘古BPM
+ * @author 分享牛
+ */
+public class FlowableTestTaskInQueryTest extends FlowableTestCase {
+
+    @Deployment
+    public void testSimpleProcess() {
```
Comment (filiphr) [inline modules/flowable-engine/src/test/java/org/flowable/standalone/testing/FlowableTestTaskInQueryTest.java:30]: Can you please move this test ot the `TaskQueryTest` we don't need another test class for this

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @qiudaoke. For some reason I missed it, sorry for the late review.

I've left some comments that would be good to be addressed before merging this.

In addition to that will you have time to expose this to the REST API as well? In the:

* `org.flowable.cmmn.rest.service.api.runtime.task.TaskCollectionResource`
* `org.flowable.cmmn.rest.service.api.history.task.HistoricTaskInstanceCollectionResource`
* `org.flowable.cmmn.rest.service.api.runtime.task.TaskQueryResource`
* `org.flowable.rest.service.api.runtime.task.TaskCollectionResource`
* `org.flowable.rest.service.api.history.HistoricTaskInstanceCollectionResource`
* `org.flowable.rest.service.api.runtime.task.TaskQueryResource`

If not we can add it separately later

**Comment by filiphr** [comment]
@qiudaoke can you please rebase the PR on top of main as well, we did some changes in the XML files and there are some conflicts now

**Comment by qiudaoke** [comment]
> @qiudaoke can you please rebase the PR on top of main as well, we did some changes in the XML files and there are some conflicts now
see this pr
https://github.com/flowable/flowable-engine/pull/3224

**Comment by filiphr** [comment]
Closing this since PR #3224

**Pull request closed without merging**